docs/src/installation/configurations.md

# Configuring the API service

Raw Data API can be setup using two configuration options. You can choose based on your convienience 
- *config.txt* : You can follow  `config.txt.sample` in dir and documentation below to set your configurations 
- *.env* : Another option is from OS Environment variable , You can export all your env variables ( They are same as you put in config without blocks ) and pass it to API , API will pick it up automatically.

## What you need to start?

- A working [Postgres](https://www.postgresql.org/) instance with [Postgis](https://postgis.net/) module enabled.
- A working [Redis](https://redis.io/) instance.

The default configuration file is an ini-style text file named `config.txt` in the project root.

## Users Table

Users table is present on ```backend/sql/users.sql``` Make sure you have it before moving forward

```
psql -a -f backend/sql/users.sql
```
& Add your admin's OSM ID as admin 

```
INSERT INTO users (osm_id, role) VALUES (1234, 1);
```

## Sections

The following sections are recognised.

- `[DB]` - For database connection information. Required.
- `[OAUTH]` - For connecting to OpenStreetMap using an OAuth2 app. Required.
- `[CELERY]` - For task queues on Redis. Required.
- `[API_CONFIG]` - API service related configuration. Required.
- `[EXPORT_UPLOAD]` - For external file hosts like S3. Optional.
- `[SENTRY]` - Sentry monitoring configuration. Optional.
- `[HDX]` - HDX Exports related configuration. Optional.

The following are the different configuration options that are accepted.

| Config option | ENVVAR | Section | Defaults | Description | Required? |
|---------------|---------|----------|--------|-------------|-----------|
| `PGHOST` | `PGHOST` | `[DB]` | _none_  | PostgreSQL hostname or IP | REQUIRED |
| `PGPORT` | `PGPORT` | `[DB]` | `5432` | PostgreSQL connection port | OPTIONAL |
| `PGUSER` | `PGUSER` | `[DB]` | _none_ | PostgreSQL user/role | REQUIRED |
| `PGPASSWORD` | `PGPASSWORD` | `[DB]` | _none_ | PostgreSQL user/role password | REQUIRED |
| `PGDATABASE` | `PGDATABASE` | `[DB]` | _none_ | PostgreSQL database name | REQUIRED |
| `OSM_CLIENT_ID` | `OSM_CLIENT_ID` | `[OAUTH]` | _none_ | Client ID of OSM OAuth2 application | REQIRED |
| `OSM_CLIENT_SECRET` | `OSM_CLIENT_SECRET` | `[OAUTH]` | _none_ | Client Secret of OSM OAuth2 application | REQIRED |
| `OSM_PERMISSION_SCOPE` | `OSM_PERMISSION_SCOPE` | `[OAUTH]` | `read_prefs` | OSM access permission for OAuth2 application | OPTIONAL |
| `LOGIN_REDIRECT_URI` | `LOGIN_REDIRECT_URI` | `[OAUTH]` | _none_ | Redirect URL set in the OAuth2 application | REQUIRED |
| `APP_SECRET_KEY` | `APP_SECRET_KEY` | `[OAUTH]` | _none_ | High-entropy string generated for the application | REQUIRED |
| `OSM_URL` | `OSM_URL` | `[OAUTH]` | `https://www.openstreetmap.org` | OSM instance Base URL | OPTIONAL |
| `LOG_LEVEL` | `LOG_LEVEL` | `[API_CONFIG]` | `debug` | Application log level; info,debug,warning,error | OPTIONAL |
| `RATE_LIMITER_STORAGE_URI` | `RATE_LIMITER_STORAGE_URI` | `[API_CONFIG]` | `redis://redis:6379` | Redis connection string for rate-limiter data | OPTIONAL |
| `RATE_LIMIT_PER_MIN` | `RATE_LIMIT_PER_MIN` | `[API_CONFIG]` | `5` | Number of requests per minute before being rate limited | OPTIONAL |
| `EXPORT_PATH` | `EXPORT_PATH` | `[API_CONFIG]` | `exports`? |  Local path to store exports | OPTIONAL |
| `EXPORT_MAX_AREA_SQKM` | `EXPORT_MAX_AREA_SQKM` | `[API_CONFIG]` | `100000` | max area in sq. km. to support for rawdata input | OPTIONAL |
| `USE_CONNECTION_POOLING` | `USE_CONNECTION_POOLING` | `[API_CONFIG]` | `false` | Enable psycopg2 connection pooling | OPTIONAL |
| `ALLOW_BIND_ZIP_FILTER` | `ALLOW_BIND_ZIP_FILTER` | `[API_CONFIG]` | `true` | Enable zip compression for exports | OPTIONAL |
| `EXTRA_README_TXT` | `EXTRA_README_TXT` | `[API_CONFIG]` | `` | Append extra string to export readme.txt | OPTIONAL |
| `ENABLE_TILES` | `ENABLE_TILES` | `[API_CONFIG]` | `false` | Enable Tile Output (Pmtiles and Mbtiles) | OPTIONAL |
| `ENABLE_SOZIP` | `ENABLE_SOZIP` | `[API_CONFIG]` | `false` | Enables sozip compression | OPTIONAL |
| `DEFAULT_QUEUE_NAME` | `DEFAULT_QUEUE_NAME` | `[API_CONFIG]` | `raw_daemon` | Option to define default queue name| OPTIONAL |
| `ONDEMAND_QUEUE_NAME` | `ONDEMAND_QUEUE_NAME` | `[API_CONFIG]` | `raw_ondemand` | Option to define daemon queue name for scheduled and long exports | OPTIONAL |
| `ENABLE_POLYGON_STATISTICS_ENDPOINTS` | `ENABLE_POLYGON_STATISTICS_ENDPOINTS` | `[API_CONFIG]` | `False` | Option to enable endpoints related the polygon statistics about the approx buildings,road length in passed polygon| OPTIONAL |
| `ENABLE_CUSTOM_EXPORTS` | `ENABLE_CUSTOM_EXPORTS` | `[API_CONFIG]` | False | Enables custom exports endpoint and imports | OPTIONAL |
| `POLYGON_STATISTICS_API_URL` | `POLYGON_STATISTICS_API_URL` | `[API_CONFIG]` | `None` | API URL for the polygon statistics to fetch the metadata , Currently tested with graphql query endpoint of Kontour , Only required if it is enabled from ENABLE_POLYGON_STATISTICS_ENDPOINTS | OPTIONAL |
| `POLYGON_STATISTICS_API_URL` | `POLYGON_STATISTICS_API_RATE_LIMIT` | `[API_CONFIG]` | `5` | Rate limit to be applied for statistics endpoint per minute, Defaults to 5 request is allowed per minute | OPTIONAL |
| `WORKER_PREFETCH_MULTIPLIER` | `WORKER_PREFETCH_MULTIPLIER` | `[CELERY]` | `1` | No of tasks that worker can prefetch at a time | OPTIONAL |
| `DEFAULT_SOFT_TASK_LIMIT` | `DEFAULT_SOFT_TASK_LIMIT` | `[API_CONFIG]` | `7200` | Soft task time limit signal for celery workers in seconds.It will gently remind celery to finish up the task and terminate, Defaults to 2 Hour| OPTIONAL |
| `DEFAULT_HARD_TASK_LIMIT` | `DEFAULT_HARD_TASK_LIMIT` | `[API_CONFIG]` | `10800` | Hard task time limit signal for celery workers in seconds. It will immediately kill the celery task.Defaults to 3 Hour| OPTIONAL |
| `USE_DUCK_DB_FOR_CUSTOM_EXPORTS` | `USE_DUCK_DB_FOR_CUSTOM_EXPORTS` | `[API_CONFIG]` | `False` | Enable this setting to use duckdb , By default duck db is disabled and postgres is used| OPTIONAL |
| `CELERY_BROKER_URL` | `CELERY_BROKER_URL` | `[CELERY]` | `redis://localhost:6379/0` | Redis connection string for the broker | OPTIONAL |
| `CELERY_RESULT_BACKEND` | `CELERY_RESULT_BACKEND` | `[CELERY]` | `redis://localhost:6379/0` | Redis/psotgresql connection string for the the result backend, eg : db+postgresql://username:password@localhost:5432/db_name  | OPTIONAL |
| `FILE_UPLOAD_METHOD` | `FILE_UPLOAD_METHOD` | `[EXPORT_UPLOAD]` | `disk` | File upload method; Allowed values - disk, s3 | OPTIONAL |
| `BUCKET_NAME` | `BUCKET_NAME` | `[EXPORT_UPLOAD]` | _none_ | AWS S3 Bucket name | CONDITIONAL |
| `AWS_ACCESS_KEY_ID` | `AWS_ACCESS_KEY_ID` | `[EXPORT_UPLOAD]` | _none_ | AWS Access Key ID for S3 access | CONDITIONAL |
| `AWS_SECRET_ACCESS_KEY` | `AWS_SECRET_ACCESS_KEY` | `[EXPORT_UPLOAD]` | _none_ | AWS Secret Access Key for S3 access | CONDITIONAL |
| `SENTRY_DSN` | `SENTRY_DSN` | `[SENTRY]` | _none_ | Sentry Data Source Name | OPTIONAL |
| `SENTRY_RATE` | `SENTRY_RATE` | `[SENTRY]` | `1.0` | Sample rate percentage for shipping errors to sentry; Allowed values between 0 (0%) to 1 (100%)| OPTIONAL |
| `ENABLE_HDX_EXPORTS` | `ENABLE_HDX_EXPORTS` | `[HDX]` | False | Enables hdx related endpoints and imports | OPTIONAL |
| `ENABLE_METRICS_APIS` | `ENABLE_METRICS_APIS` | `[API_CONFIG]` | False | Enables download metrics related endpoints , Require different setup of metrics populator | OPTIONAL |
| `HDX_SITE` | `HDX_SITE` | `[HDX]` | 'demo' | HDX site to point , By default demo site , use prod for production | CONDITIONAL |
| `HDX_API_KEY` | `HDX_API_KEY` | `[HDX]` | None | Your API Secret key for hdx upload , should have write access and it is compulsory if ENABLE_HDX_EXPORTS is True | CONDITIONAL |
| `HDX_OWNER_ORG` | `HDX_OWNER_ORG` | `[HDX]` | None | Your HDX organization ID| CONDITIONAL |
| `HDX_MAINTAINER` | `HDX_MAINTAINER` | `[HDX]` | None | Your HDX Maintainer ID | CONDITIONAL |
| `DUCK_DB_MEMORY_LIMIT` | `DUCK_DB_MEMORY_LIMIT` | `[API_CONFIG]` | None | Duck DB max memory limit , 80 % of your RAM eg : '5GB'| CONDITIONAL |

CHANGELOG.md
## 1.5.1 (2024-08-20)

### Fix

- **staff-access-for-download-metrics**: fixes bug for metrics not being available to staff

## 1.5.0 (2024-08-19)

### Feat

- **download-metrics**: enabels user to fetch and download metrics for rawdataapi

## 1.4.6 (2024-08-09)

### Fix

- **unit-test-failing**: only add file format if tiles are enabled

### Perf

- **test-cases**: don't run if task is failed

## 1.4.5 (2024-08-09)

### Fix

- **mvt-tiles**: fixes zoom level and zipping method for mvt tiles

## 1.4.4 (2024-08-09)

### Fix

- **security-patch-task-result**: don't raise the error to user when task status is being returned

## 1.4.3 (2024-08-08)

### Fix

- **mvt-format**: fix bug on mvt formats invalid command line args

## 1.4.2 (2024-08-08)

### Fix

- **fix-zoom-level-bug-on-pmtiles**: adds proper syntax on vector tiles exports zoom level

## 1.4.1 (2024-08-08)

### Fix

- **change-default-db-user-to-user-column**: default user was being used for db query instead read it from table itself
- **fix-on-dir-size-issue**: failing exports causing dir to explode

### Refactor

- **init-default-values**: reinitialize the default value as other endpoints don't require parameters

## 1.4.0 (2024-08-08)

### Feat

- **streaming-response-for-api**: adds streaming response for current snapshot plain
- **userinfo-on-exports**: adds userinfo in exports only for logged in users
- **vector-tiles**: support for vector tiles , allow user to select all tags in select
- **dockerfile**: minimizes docker image size  and introduces new version of gdal which should enable the latest vector driver support
- **custom_exports-yaml**: adds a yaml endpoint for custom exports

### Fix

- **fixes-failing-test-cases**: moves user checking to base class
- **wild-card-select**: fixes bug on wild card select for the query builder

### Refactor

- **cleanup-previous-non-async-code**: removes non async code for the quick fetch
- **yaml-models**: refactor model to models.py
- **custom-exports-yaml**: added geometry and both yaml validation

### Perf

- **custom-exports-yaml**: moves geometry to request body within yaml itself instead of query param

## 1.3.0 (2024-06-06)

### Feat

- **polygon-stats**: let admin and staff bypass the area limit

### Fix

- **iso3-stats**: fix bug on iso3 stats after the role check

## 1.2.2 (2024-05-02)

### Fix

- **api_worker**: clean dir when task fails

## 1.2.1 (2024-04-29)

### Fix

- **hdx-lib**: upgraded the hdx lib to fix last-modified bug

## 1.2.0 (2024-04-05)

### Feat

- **countries**: adds cid country get endpoint

## 1.1.3 (2024-04-05)

### Fix

- **test_api**: unittest fix for plain endpoint

## 1.1.2 (2024-04-04)

### Perf

- **s3downloadlink**: changes download link to support ipv6

## 1.1.1 (2024-03-29)

### Fix

- **stats**: adds area threshold in the stats api along with timeout reduced to 30 second

## 1.1.0 (2024-03-28)

### Feat

- **version-control**: Allows rawdataapi to follow version control using commitizen

### Fix

- **dokcer-compose**: One command up docker compose method
- **dockerfile**: Avoids copying tippicone for source code level changes

### Refactor

- **logging**: adds tqdm for better logging and upgrades hdx python lib
- **queue-name**: changes queue name and workers to pick correct variable
- **api-worker**: logic changes for memory optimized zip

### Perf

- **workers**: enables max worker variable to be set from env variable

## 1.0.17 (2024-03-06)

### Fix

- add optional param to wrap flatgeobuf in geomcollection
- default add spatial index to generated flatgeobuf files

### Refactor

- remove ENCODING param for flatgeobuf (does nothing)
- use f-string for flatgeobuf ogr command

## 1.0.16 (2022-12-19)

## 1.0.15 (2022-12-13)

## 1.0.14 (2022-12-13)

## 1.0.13 (2022-12-07)

## 1.0.12 (2022-12-07)

## 1.0.11 (2022-12-07)

## 1.0.10 (2022-12-07)

## 1.0.9 (2022-12-07)

## 1.0.8 (2022-12-07)

## 1.0.7 (2022-12-04)

## v1.0.6 (2022-09-15)

## v1.0.5 (2022-09-09)

## v1.0.4 (2022-08-02)

## v1.0.3 (2022-08-01)

## v1.0.2 (2022-08-01)

## v1.0.1 (2022-07-29)


src/config.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Copyright (C) 2021 Humanitarian OpenStreetmap Team

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.

# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

# Humanitarian OpenStreetmap Team
# 1100 13th Street NW Suite 800 Washington, D.C. 20005
# <info@hotosm.org>

# Standard library imports
import logging
import os
from configparser import ConfigParser
from distutils.util import strtobool

# Third party imports
from slowapi import Limiter
from slowapi.util import get_remote_address


def get_bool_env_var(key, default=False):
    value = os.environ.get(key, default)
    return bool(strtobool(str(value)))


CONFIG_FILE_PATH = "config.txt"
USE_S3_TO_UPLOAD = False
AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY, BUCKET_NAME = None, None, None


config = ConfigParser()
config.read(CONFIG_FILE_PATH)


### CELERY BLOCK ####################
CELERY_BROKER_URL = os.environ.get("CELERY_BROKER_URL") or config.get(
    "CELERY", "CELERY_BROKER_URL", fallback="redis://localhost:6379"
)
CELERY_RESULT_BACKEND = os.environ.get("CELERY_RESULT_BACKEND") or config.get(
    "CELERY", "CELERY_RESULT_BACKEND", fallback="redis://localhost:6379"
)

WORKER_PREFETCH_MULTIPLIER = int(
    os.environ.get("WORKER_PREFETCH_MULTIPLIER")
    or config.get("CELERY", "WORKER_PREFETCH_MULTIPLIER", fallback=1)
)

### API CONFIG BLOCK #######################

RATE_LIMIT_PER_MIN = os.environ.get("RATE_LIMIT_PER_MIN") or int(
    config.get("API_CONFIG", "RATE_LIMIT_PER_MIN", fallback=20)
)

RATE_LIMITER_STORAGE_URI = os.environ.get("RATE_LIMITER_STORAGE_URI") or config.get(
    "API_CONFIG", "RATE_LIMITER_STORAGE_URI", fallback="redis://localhost:6379"
)

EXPORT_MAX_AREA_SQKM = os.environ.get("EXPORT_MAX_AREA_SQKM") or int(
    config.get("API_CONFIG", "EXPORT_MAX_AREA_SQKM", fallback=100000)
)


INDEX_THRESHOLD = os.environ.get("INDEX_THRESHOLD") or int(
    config.get("API_CONFIG", "INDEX_THRESHOLD", fallback=5000)
)

MAX_WORKERS = os.environ.get("MAX_WORKERS") or config.get(
    "API_CONFIG", "MAX_WORKERS", fallback=os.cpu_count()
)

# get log level from config
LOG_LEVEL = os.environ.get("LOG_LEVEL") or config.get(
    "API_CONFIG", "LOG_LEVEL", fallback="debug"
)


def not_raises(func, *args, **kwargs):
    try:
        func(*args, **kwargs)
        return True
    except Exception as ex:
        logging.error(ex)
        return False


####################

# EXPORT_UPLOAD CONFIG BLOCK
FILE_UPLOAD_METHOD = os.environ.get("FILE_UPLOAD_METHOD") or config.get(
    "EXPORT_UPLOAD", "FILE_UPLOAD_METHOD", fallback="disk"
)


if FILE_UPLOAD_METHOD.lower() not in ["s3", "disk"]:
    logging.error(
        "value not supported for file_upload_method ,switching to default disk method"
    )
    USE_S3_TO_UPLOAD = False

if FILE_UPLOAD_METHOD.lower() == "s3":
    USE_S3_TO_UPLOAD = True
    BUCKET_NAME = os.environ.get("BUCKET_NAME") or config.get(
        "EXPORT_UPLOAD", "BUCKET_NAME"
    )
    AWS_ACCESS_KEY_ID = os.environ.get("AWS_ACCESS_KEY_ID") or config.get(
        "EXPORT_UPLOAD", "AWS_ACCESS_KEY_ID", fallback=None
    )
    AWS_SECRET_ACCESS_KEY = os.environ.get("AWS_SECRET_ACCESS_KEY") or config.get(
        "EXPORT_UPLOAD", "AWS_SECRET_ACCESS_KEY", fallback=None
    )
    if not BUCKET_NAME:
        raise ValueError("Value of BUCKET_NAME couldn't found")

##################

## SENTRY BLOCK ########
SENTRY_DSN = os.environ.get("SENTRY_DSN") or config.get(
    "SENTRY", "SENTRY_DSN", fallback=None
)
SENTRY_RATE = os.environ.get("SENTRY_RATE") or config.get(
    "SENTRY", "SENTRY_RATE", fallback=None
)

# rate limiter for API requests based on the remote ip address and redis as backend
LIMITER = Limiter(key_func=get_remote_address, storage_uri=RATE_LIMITER_STORAGE_URI)


if LOG_LEVEL.lower() == "debug":  # default debug
    level = logging.DEBUG
elif LOG_LEVEL.lower() == "info":
    level = logging.INFO
elif LOG_LEVEL.lower() == "error":
    level = logging.ERROR
elif LOG_LEVEL.lower() == "warning":
    level = logging.WARNING
else:
    logging.error(
        "logging config is not supported , Supported fields are : debug,error,warning,info , Logging to default :debug"
    )
    level = logging.DEBUG

# logging.getLogger("fiona").propagate = False  # disable fiona logging
logging.basicConfig(format="%(asctime)s - %(message)s", level=level)
logging.getLogger("boto3").propagate = False  # disable boto3 logging
logging.getLogger("botocore").propagate = False  # disable boto3 logging
logging.getLogger("s3transfer").propagate = False  # disable boto3 logging
logging.getLogger("boto").propagate = False  # disable boto3 logging


## Readme txt


logger = logging.getLogger("raw_data_api")

EXPORT_PATH = os.environ.get("EXPORT_PATH") or config.get(
    "API_CONFIG", "EXPORT_PATH", fallback="exports"
)

if not os.path.exists(EXPORT_PATH):
    # Create a exports directory because it does not exist
    os.makedirs(EXPORT_PATH)


DEFAULT_README_TEXT = """Exported through Raw-data-api (https://github.com/hotosm/raw-data-api) using OpenStreetMap data.\n Exports are made available under the Open Database License: http://opendatacommons.org/licenses/odbl/1.0/. Any rights in individual contents of the database are licensed under the Database Contents License: http://opendatacommons.org/licenses/dbcl/1.0/. \n Learn more about OpenStreetMap and its data usage policy : https://www.openstreetmap.org/about \n"""

EXTRA_README_TXT = os.environ.get("EXTRA_README_TXT") or config.get(
    "API_CONFIG", "EXTRA_README_TXT", fallback=""
)
DEFAULT_README_TEXT += EXTRA_README_TXT

ALLOW_BIND_ZIP_FILTER = get_bool_env_var(
    "ALLOW_BIND_ZIP_FILTER",
    config.getboolean("API_CONFIG", "ALLOW_BIND_ZIP_FILTER", fallback=False),
)

ENABLE_SOZIP = get_bool_env_var(
    "ENABLE_SOZIP",
    config.getboolean("API_CONFIG", "ENABLE_SOZIP", fallback=False),
)

ENABLE_TILES = get_bool_env_var(
    "ENABLE_TILES", config.getboolean("API_CONFIG", "ENABLE_TILES", fallback=False)
)

# check either to use connection pooling or not
USE_CONNECTION_POOLING = get_bool_env_var(
    "USE_CONNECTION_POOLING",
    config.getboolean("API_CONFIG", "USE_CONNECTION_POOLING", fallback=False),
)

# Queue

DEFAULT_QUEUE_NAME = os.environ.get("DEFAULT_QUEUE_NAME") or config.get(
    "API_CONFIG", "DEFAULT_QUEUE_NAME", fallback="raw_daemon"
)
ONDEMAND_QUEUE_NAME = os.environ.get("ONDEMAND_QUEUE_NAME") or config.get(
    "API_CONFIG", "ONDEMAND_QUEUE_NAME", fallback="raw_ondemand"
)

# Polygon statistics which will deliver the stats of approx buildings/ roads in the area

ENABLE_POLYGON_STATISTICS_ENDPOINTS = get_bool_env_var(
    "ENABLE_POLYGON_STATISTICS_ENDPOINTS",
    config.getboolean(
        "API_CONFIG", "ENABLE_POLYGON_STATISTICS_ENDPOINTS", fallback=False
    ),
)
POLYGON_STATISTICS_API_URL = os.environ.get("POLYGON_STATISTICS_API_URL") or config.get(
    "API_CONFIG", "POLYGON_STATISTICS_API_URL", fallback=None
)

POLYGON_STATISTICS_API_RATE_LIMIT = os.environ.get(
    "POLYGON_STATISTICS_API_RATE_LIMIT"
) or config.get("API_CONFIG", "POLYGON_STATISTICS_API_RATE_LIMIT", fallback=5)

# task limit

DEFAULT_SOFT_TASK_LIMIT = os.environ.get("DEFAULT_SOFT_TASK_LIMIT") or config.get(
    "API_CONFIG", "DEFAULT_SOFT_TASK_LIMIT", fallback=2 * 60 * 60
)
DEFAULT_HARD_TASK_LIMIT = os.environ.get("DEFAULT_HARD_TASK_LIMIT") or config.get(
    "API_CONFIG", "DEFAULT_HARD_TASK_LIMIT", fallback=3 * 60 * 60
)

# duckdb

USE_DUCK_DB_FOR_CUSTOM_EXPORTS = get_bool_env_var(
    "USE_DUCK_DB_FOR_CUSTOM_EXPORTS",
    config.getboolean("API_CONFIG", "USE_DUCK_DB_FOR_CUSTOM_EXPORTS", fallback=False),
)

logger.info(
    "USE_DUCK_DB_FOR_CUSTOM_EXPORTS %s ", USE_DUCK_DB_FOR_CUSTOM_EXPORTS is True
)

if USE_DUCK_DB_FOR_CUSTOM_EXPORTS:
    DUCK_DB_MEMORY_LIMIT = os.environ.get("DUCK_DB_MEMORY_LIMIT") or config.get(
        "API_CONFIG", "DUCK_DB_MEMORY_LIMIT", fallback=None
    )
    DUCK_DB_THREAD_LIMIT = os.environ.get("DUCK_DB_THREAD_LIMIT") or config.get(
        "API_CONFIG", "DUCK_DB_THREAD_LIMIT", fallback=None
    )

# hdx and custom exports
ENABLE_CUSTOM_EXPORTS = get_bool_env_var(
    "ENABLE_CUSTOM_EXPORTS",
    config.getboolean("API_CONFIG", "ENABLE_CUSTOM_EXPORTS", fallback=False),
)

HDX_SOFT_TASK_LIMIT = os.environ.get("HDX_SOFT_TASK_LIMIT") or config.get(
    "HDX", "HDX_SOFT_TASK_LIMIT", fallback=5 * 60 * 60
)
HDX_HARD_TASK_LIMIT = os.environ.get("HDX_HARD_TASK_LIMIT") or config.get(
    "HDX", "HDX_HARD_TASK_LIMIT", fallback=6 * 60 * 60
)

ENABLE_HDX_EXPORTS = get_bool_env_var(
    "ENABLE_HDX_EXPORTS", config.getboolean("HDX", "ENABLE_HDX_EXPORTS", fallback=False)
)

PROCESS_SINGLE_CATEGORY_IN_POSTGRES = get_bool_env_var(
    "PROCESS_SINGLE_CATEGORY_IN_POSTGRES",
    config.getboolean("HDX", "PROCESS_SINGLE_CATEGORY_IN_POSTGRES", fallback=False),
)

PARALLEL_PROCESSING_CATEGORIES = get_bool_env_var(
    "PARALLEL_PROCESSING_CATEGORIES",
    config.getboolean("HDX", "PARALLEL_PROCESSING_CATEGORIES", fallback=True),
)


ENABLE_METRICS_APIS = get_bool_env_var(
    "ENABLE_METRICS_APIS",
    config.getboolean("API_CONFIG", "ENABLE_METRICS_APIS", fallback=False),
)


if ENABLE_HDX_EXPORTS:
    try:
        hdx_credentials = os.environ["REMOTE_HDX"]

    except KeyError:
        logger.debug("EnvVar: REMOTE_HDX not supplied; Falling back to other means")
        HDX_SITE = os.environ.get("HDX_SITE") or config.get(
            "HDX", "HDX_SITE", fallback="demo"
        )
        HDX_API_KEY = os.environ.get("HDX_API_KEY") or config.get(
            "HDX", "HDX_API_KEY", fallback=None
        )
        HDX_OWNER_ORG = os.environ.get("HDX_OWNER_ORG") or config.get(
            "HDX", "HDX_OWNER_ORG", fallback="225b9f7d-e7cb-4156-96a6-44c9c58d31e3"
        )
        HDX_MAINTAINER = os.environ.get("HDX_MAINTAINER") or config.get(
            "HDX", "HDX_MAINTAINER", fallback=None
        )

    else:
        # Standard library imports
        import json

        hdx_credentials_json = json.loads(hdx_credentials)

        HDX_SITE = hdx_credentials_json["HDX_SITE"]
        HDX_API_KEY = hdx_credentials_json["HDX_API_KEY"]
        HDX_OWNER_ORG = hdx_credentials_json["HDX_OWNER_ORG"]
        HDX_MAINTAINER = hdx_credentials_json["HDX_MAINTAINER"]

        if None in (HDX_SITE, HDX_API_KEY, HDX_OWNER_ORG, HDX_MAINTAINER):
            raise ValueError("HDX Remote Credentials Malformed")

    # Third party imports
    from hdx.api.configuration import Configuration

    try:
        HDX_URL_PREFIX = Configuration.create(
            hdx_site=HDX_SITE,
            hdx_key=HDX_API_KEY,
            user_agent="HDXPythonLibrary/6.2.0-HOTOSM OSM Exports",
        )
        logging.debug(HDX_URL_PREFIX)
    except Exception as e:
        logging.error(
            "Error creating HDX configuration: %s, Disabling the hdx exports feature", e
        )
        ENABLE_HDX_EXPORTS = False

if ENABLE_HDX_EXPORTS:
    # Third party imports
    from hdx.data.dataset import Dataset
    from hdx.data.vocabulary import Vocabulary

    parse_list = lambda value, delimiter=",": (
        value.split(delimiter) if isinstance(value, str) else value or []
    )

    ALLOWED_HDX_TAGS = parse_list(
        os.environ.get("ALLOWED_HDX_TAGS")
        or config.get("HDX", "ALLOWED_HDX_TAGS", fallback=None)
        or (
            Vocabulary.approved_tags() if not_raises(Vocabulary.approved_tags) else None
        )
    )
    ALLOWED_HDX_UPDATE_FREQUENCIES = parse_list(
        os.environ.get("ALLOWED_HDX_UPDATE_FREQUENCIES")
        or config.get("HDX", "ALLOWED_HDX_UPDATE_FREQUENCIES", fallback=None)
        or (
            Dataset.list_valid_update_frequencies()
            if not_raises(Dataset.list_valid_update_frequencies)
            else None
        )
    )


def get_db_connection_params() -> dict:
    """Return a python dict that can be passed to psycopg2 connections
    to authenticate to Postgres Databases

    Returns: connection_params (dict): PostgreSQL connection parameters
             corresponding to the configuration section.

    """
    # This block fetches PostgreSQL (database) credentials passed as
    # environment variables as a JSON object, from AWS Secrets Manager or
    # Azure Key Vault.
    try:
        db_credentials = os.environ["REMOTE_DB"]

    except KeyError:
        logger.debug("EnvVar: REMOTE_DB not supplied; Falling back to other means")

        connection_params = dict(
            host=os.environ.get("PGHOST") or config.get("DB", "PGHOST"),
            port=os.environ.get("PGPORT")
            or config.get("DB", "PGPORT", fallback="5432"),
            dbname=os.environ.get("PGDATABASE") or config.get("DB", "PGDATABASE"),
            user=os.environ.get("PGUSER") or config.get("DB", "PGUSER"),
            password=os.environ.get("PGPASSWORD") or config.get("DB", "PGPASSWORD"),
        )

    else:
        # Standard library imports
        import json

        connection_params = json.loads(db_credentials)

        connection_params["user"] = connection_params["username"]
        for k in ("dbinstanceidentifier", "dbClusterIdentifier", "engine", "username"):
            if k in connection_params:
                connection_params.pop(k, None)

    if None in connection_params.values():
        raise ValueError(
            "Connection Params Value Error :  Couldn't be Loaded , Check DB Credentials"
        )
        logging.error(
            "Can't find database credentials , Either export them as env variable or include in config Block DB"
        )

    return connection_params


def get_oauth_credentials() -> tuple:
    """Get OAuth2 credentials from env file and return a config dict

    Return an ordered python tuple that can be passed to functions that
    authenticate to OSM.

    Order of precedence:
    1. Environment Variables
    2. Config File
    3. Default fallback

    Returns: oauth2_credentials (tuple): Tuple containing OAuth2 client
             secret, client ID, and redirect URL.

    """
    # This block fetches OSM OAuth2 app credentials passed as
    # environment variables as a JSON object, from AWS Secrets Manager or
    # Azure Key Vault.
    osm_url = os.environ.get("OSM_URL") or config.get(
        "OAUTH", "OSM_URL", fallback="https://www.openstreetmap.org"
    )
    secret_key = os.environ.get("APP_SECRET_KEY") or config.get(
        "OAUTH", "APP_SECRET_KEY"
    )

    try:
        oauth2_credentials = os.environ["REMOTE_OAUTH"]
    except KeyError:
        logger.debug("EnvVar: REMOTE_OAUTH not supplied; Falling back to other means")

        client_id = os.environ.get("OSM_CLIENT_ID") or config.get(
            "OAUTH", "OSM_CLIENT_ID"
        )
        client_secret = os.environ.get("OSM_CLIENT_SECRET") or config.get(
            "OAUTH", "OSM_CLIENT_SECRET"
        )
        login_redirect_uri = os.environ.get("LOGIN_REDIRECT_URI") or config.get(
            "OAUTH",
            "LOGIN_REDIRECT_URI",
            fallback="http://127.0.0.1:8000/v1/auth/callback",
        )
        scope = os.environ.get("OSM_PERMISSION_SCOPE") or config.get(
            "OAUTH", "OSM_PERMISSION_SCOPE", fallback="read_prefs"
        )

    else:
        # Standard library imports
        import json

        oauth2_credentials_json = json.loads(oauth2_credentials)

        client_id = oauth2_credentials_json["OSM_CLIENT_ID"]
        client_secret = oauth2_credentials_json["OSM_CLIENT_SECRET"]
        login_redirect_uri = oauth2_credentials_json["LOGIN_REDIRECT_URI"]
        scope = oauth2_credentials_json["OSM_PERMISSION_SCOPE"]

    oauth_cred = (
        osm_url,
        client_id,
        client_secret,
        secret_key,
        login_redirect_uri,
        scope,
    )

    if None in oauth_cred:
        raise ValueError("Oauth Credentials can't be loaded")

    return oauth_cred
